Source/Flow/Private/FlowAsset.cpp

```diff
@@ -421,34 +470,76 @@ void UFlowAsset::FinishFlow(const EFlowFinishPolicy InFinishPolicy, const bool b
 	}
 }
 
+bool UFlowAsset::HasStartedFlow() const
+{
+	return RecordedNodes.Num() > 0;
+}
+
+AActor* UFlowAsset::TryFindActorOwner() const
+{
+	UActorComponent* OwnerAsComponent = Cast<UActorComponent>(GetOwner());
+	if (IsValid(OwnerAsComponent))
+	{
+		return Cast<AActor>(OwnerAsComponent->GetOwner());
+	}
+
+	return nullptr;
+}
+
 TWeakObjectPtr<UFlowAsset> UFlowAsset::GetFlowInstance(UFlowNode_SubGraph* SubGraphNode) const
 {
 	return ActiveSubGraphs.FindRef(SubGraphNode);
 }
 
-void UFlowAsset::TriggerCustomEvent(UFlowNode_SubGraph* Node, const FName& EventName) const
+void UFlowAsset::TriggerSubgraphCustomInput(UFlowNode_SubGraph& Node, const FName& EventName) const
 {
-	const TWeakObjectPtr<UFlowAsset> FlowInstance = ActiveSubGraphs.FindRef(Node);
+	check(HasStartedFlow());
+
+	const TWeakObjectPtr<UFlowAsset> FlowInstance = ActiveSubGraphs.FindRef(&Node);
 	if (FlowInstance.IsValid())
 	{
-		for (UFlowNode_CustomInput* CustomInput : FlowInstance->CustomInputNodes)
+		FlowInstance->TriggerCustomInput(EventName);
+	}
+}
+
+void UFlowAsset::TriggerCustomInput(const FName& EventName)
+{
+	check(HasStartedFlow());
+
+	for (UFlowNode_CustomInput* CustomInput : CustomInputNodes)
+	{
+		if (CustomInput->EventName == EventName)
 		{
-			if (CustomInput->EventName == EventName)
-			{
-				FlowInstance->RecordedNodes.Add(CustomInput);
-				CustomInput->TriggerFirstOutput(true);
-			}
+			RecordedNodes.Add(CustomInput);
+
+			CustomInput->ExecuteInput(EventName);
 		}
 	}
 }
 
-void UFlowAsset::TriggerCustomOutput(const FName& EventName) const
+void UFlowAsset::TriggerCustomOutput(const FName& EventName)
 {
-	NodeOwningThisAssetInstance->TriggerOutput(EventName);
+	check(HasStartedFlow());
+
+	const bool bIsSubgraph = NodeOwningThisAssetInstance.IsValid();
+	if (bIsSubgraph)
+	{
+		NodeOwningThisAssetInstance->TriggerOutput(EventName);
+	}
+	else
+	{
+		// Non-subgraphs are Root instances, so call the  OnTriggerCustomOutputEventDispatcher
+		if (UFlowComponent* FlowComponent = Cast<UFlowComponent>(GetOwner()))
+		{
+			FlowComponent->OnTriggerRootFlowOutputEventDispatcher(*this, EventName);
+		}
+	}
 }
 
 void UFlowAsset::TriggerInput(const FGuid& NodeGuid, const FName& PinName)
 {
+	check(HasStartedFlow());
```

Comment:
I wanted to catch any cases where TriggerInput() was called prior to StartFlow() being called.

Reply:
I'm sorry, but I'm removing all these checks which might crash in the editor (leading to lost work of a given developer) or even in released games in some edge cases.
I'd recommend using here at least logging which would inform a non-programmer what happened. I spent too much time in life on resolving crash reports caused by issued that might easily fixed by designer alone :)

I understand the need of adding checks everywhere to catch invalid data/events during development. Might accept `check` in runtime code, if it was optionally compiled out :)

Reply:
in UE stock, "check" is not included in Test or Shipping configurations.  "ensure" is included in all versions.

[https://docs.unrealengine.com/4.26/en-US/ProgrammingAndScripting/ProgrammingWithCPP/Assertions/](url)

For my team, we use check() for programmer errors, to require that programmers are calling the code as expected.  For any errors that can be caused by blueprint or non-programmers, we use the unreal error reporting tools (often MessageLog or UE_LOG).  This sort of check is appropriate according to my code standard, because it polices **code** calling the functions incorrectly, and would be found and fixed by a programmer before ever being put in front of an artist or designer's eyes.

All that said, I don't mind you stripping them out here, it's ultimately your code.